Further API-capabilities for plugins

NobodysNightmare opened the pull request:
## OpenProject Work Package

https://community.openproject.org/work_packages/19044
## Summary

This PR adds further capabilities required by https://github.com/finnlabs/openproject-backlogs/pull/123.

NobodysNightmare committed:
allow to patch classes in a namespace

NobodysNightmare committed:
add type getter to WP schema

NobodysNightmare committed:
allow passing of callables to writable + required

NobodysNightmare committed:
"improve" work package contract

trying to put related validations closer together:
- add writable properties using a method call
- additional validations for that property take place in the DO block

also allows for better extension by plugins

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-even-more-capabilities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-even-more-capabilities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented on lib/api/decorators/schema.rb:
```diff
@@ -48,8 +48,8 @@ def schema(property,
                                                                                name: name,
-                                                                               required: required,
-                                                                               writable: writable)
+                                                                               required: call_or_use(required),
```
I hate that this line got this long, but it is basically only this darn namespace prefix...

I'd rather keep it as long if there are no objections...

NobodysNightmare replied:
@Azure7111 told me how to do best... I agree ^^

NobodysNightmare committed:
style

ulferts commented on lib/api/v3/work_packages/work_package_contract.rb:
```diff
@@ -27,28 +27,39 @@
-            fixed_version_id
-          )
+      class WorkPackageContract < ::API::Contracts::ModelContract
+        writable_attribute :subject
```
just a naming suggesting: we only have `writable_attributes` - a whitelist. Every attribute not listed is not writeable. Can't we just shorten this to `attribute`. That would make it easier for me to understand the validation block as I would no longer think that the block has something to do with the writing part.

ulferts replied:
Additionally, we could allow for specifying multiple attributes at the same time.

NobodysNightmare replied:
Sounds good to me,

however I'd rather not use that feature too extensively, because I feel that the readability does not increase very much, when you put a bunch of unrelated attributes into one line.

ulferts commented on lib/open_project/plugins/acts_as_op_engine.rb:
```diff
@@ -68,6 +68,18 @@ def self.included(base)
+          klass = qualified_class_name.to_s.constantize
+          klass.send(:include, patch) unless klass.included_modules.include?(patch)
+        end
+      end
```
You can get the patches constant easier and more robust like this:

```
plugin_module = self.class.to_s.deconstantize
base.config.to_prepare do
  klass_name = args.last
  patch = "#{plugin_module}::Patches::#{klass_name}Patch".constantize
  ....
end
```

And while you are at it: **you should not copy code**

NobodysNightmare replied:
I thought this is how programming works?!?

![copy & paste](http://blog.intrapromote.com/wp-content/uploads/2012/07/copy-paste-300x244.jpeg)

ulferts commented:
Functionality looks good, should be possible to merge this, soon.

NobodysNightmare committed:
improve on contract syntax

- use the shorter name "attribute"
- allow to specify multiple attributes at once

NobodysNightmare committed:
use more concise and foolproof patch-method

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/dev' into feature/even_more_capabilities

Conflicts:
	lib/api/v3/work_packages/work_package_contract.rb

ulferts commented:
@NobodysNightmare rebasing required. 
![image](https://cloud.githubusercontent.com/assets/617519/6664228/4b8185aa-cbce-11e4-864c-6e4c35113d47.png)

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/dev' into feature/even_more_capabilities

Conflicts:
	lib/api/v3/work_packages/work_package_contract.rb

NobodysNightmare commented:
Travis says it is green:

![green as the sun](http://www.tapeto.com/media/images/org/03120.jpg)

ulferts commented:
@NobodysNightmare <- Windows user.
